File: dart/lib/src/data_sources/inline_calendar_data_source.dart
import '../types.dart';

/// Data source backed by an in-memory map.
///
/// Ideal for Flutter apps (load from assets), server Dart, or tests.
///
/// Usage:
/// ```dart
/// final source = InlineCalendarDataSource(
///   years: {2082: [31,31,32,31,31,31,30,29,30,29,30,30]},
///   startYear: 2082,
///   endYear: 2082,
/// );
/// NepaliCalendar.initialize(source);
/// ```
class InlineCalendarDataSource implements NepaliCalendarDataSource {
  final Map<int, List<int>> _years;
  @override final int supportedStartYear;
  @override final int supportedEndYear;

  const InlineCalendarDataSource({
    required Map<int, List<int>> years,
    required this.supportedStartYear,
    required this.supportedEndYear,
  }) : _years = years;

  @override
  List<int>? getYearData(int year) => _years[year];
}

File: ts/src/data_sources/InlineCalendarDataSource.ts
import type { NepaliCalendarDataSource } from '../types.js';

/**
 * A data source backed by a plain JS object / Record.
 * Ideal for Node.js, server-side rendering, or test environments where
 * you already have the year data in memory.
 *
 * Usage:
 *   import data from '../../data/nepali_calendar_data.json' assert { type: 'json' };
 *   const source = new InlineCalendarDataSource(data.years, data.supportedRange.start, data.supportedRange.end);
 *
 * Or with a subset for tests:
 *   const source = new InlineCalendarDataSource({ 2082: [31,31,32,...] }, 2082, 2082);
 */
export class InlineCalendarDataSource implements NepaliCalendarDataSource {
  private readonly _yearData: Map<number, number[]>;
  private readonly _startYear: number;
  private readonly _endYear: number;

  constructor(years: Record<string, number[]>, startYear: number, endYear: number) {
    this._yearData = new Map(
      Object.entries(years).map(([k, v]) => [Number(k), v])
    );
    this._startYear = startYear;
    this._endYear = endYear;
  }

  getYearData(year: number): number[] | null {
    return this._yearData.get(year) ?? null;
  }

  get supportedStartYear(): number {
    return this._startYear;
  }

  get supportedEndYear(): number {
    return this._endYear;
  }
}
